### DISCUSSION TITLE
Clarification of DatasetGraph contract

### DISCUSSION THREAD

**Discussion by cygri** [apache/jena#1308] [Q&A] [answered]
I have some questions on the contract for `DatasetGraph`. What exactly is expected of an implementation of this interface? Let's say I want the implementation to work at least with ARQ's implementations of SPARQL Query, SPARQL Update, and the various dataset readers and writers.

- Can `getDefaultGraph()` return `null`? Can it throw `UnsupportedOperationException`?
- Can `getUnionGraph()` return `null`? Can it throw `UnsupportedOperationException`?
- Many (all?) of the provided implementations of `DatasetGraph` support the special graph names `urn:x-arq:DefaultGraph`, `urn:x-arq:DefaultGraphNode`, and `urn:x-arq:UnionGraph`. Is this required for all implementations of `DatasetGraph`?
- Is `add(null, s, p, o)` and `remove(null, s, p, o)` a valid way of adding and removing a triple from the default graph? (`DatasetGraphCollection` seems to support this usage, but I'm not sure it works with other implementations.)
- Is the following interpretation correct? The only difference between `find(g, s, p, o)` and `findNG(g, s, p, o)` is that a wildcard graph (`null` or `Node.ANY`) will also search the default graph in the former, but not in the latter. There is no difference between the methods with regards to the interpretation of the special graph names `urn:x-arq:DefaultGraph`, `urn:x-arq:DefaultGraphNode` and `urn:x-arq:UnionGraph`.
- In implementations that distinguish between empty and absent named graphs, would `dsg.isEmpty()` mean “all graphs are empty”, or “there are no named graphs”? In other words, after calling `addGraph(g, anEmptyGraph)` on an empty `DatasetGraph`, would an implementation be allowed to return `false`? Symmetry with `size()` and `clear()` would suggest that the answer is “yes”.

Thank you!

**Comment by afs** [comment, marked as answer]
> - Can `getDefaultGraph()` return `null`? Can it throw `UnsupportedOperationException`?

There is always a default graph. This is defined in SPARQL - a dataset is a default graph and zero or more named graphs.

> - Can `getUnionGraph()` return `null`? Can it throw `UnsupportedOperationException`?

There is a general purpose (and possibly inefficient) implement provided in `DatasetGraphBase`.

It is better to provide something (tests pass) but sometimes it might be meaning less and `UnsupportedOperationException` - with javadoc to explain why - would be appropriate.

> - Many (all?) of the provided implementations of `DatasetGraph` support the special graph names `urn:x-arq:DefaultGraph`, `urn:x-arq:DefaultGraphNode`, and `urn:x-arq:UnionGraph`. Is this required for all implementations of `DatasetGraph`?

The `DatasetGraph` hierarchy (a tree of various styles of implementation) uses these - directly implementing `DatasetGraph` shouldn't necessary.

> - Is `add(null, s, p, o)` and `remove(null, s, p, o)` a valid way of adding and removing a triple from the default graph? (`DatasetGraphCollection` seems to support this usage, but I'm not sure it works with other implementations.)

Yes.

> - Is the following interpretation correct? The only difference between `find(g, s, p, o)` and `findNG(g, s, p, o)` is that a wildcard graph (`null` or `Node.ANY`) will also search the default graph in the former, but not in the latter. There is no difference between the methods with regards to the interpretation of the special graph names `urn:x-arq:DefaultGraph`, `urn:x-arq:DefaultGraphNode` and `urn:x-arq:UnionGraph`.

`find` includes the default graph , `findNG` does not.

> - In implementations that distinguish between empty and absent named graphs, would `dsg.isEmpty()` mean “all graphs are empty”, or “there are no named graphs”? In other words, after calling `addGraph(g, anEmptyGraph)` on an empty `DatasetGraph`, would an implementation be allowed to return `false`? Symmetry with `size()` and `clear()` would suggest that the answer is “yes”.

All graphs are empty including the default graph.

Few implementations (just one now for legacy?) distinguish between empty and absent named graphs and this should not be relied upon. Best to plan around empty graphs not being preserved.

**Reply by afs** [reply]
Assuming the response 7/7/2022 answers the question.
